File: references/betterlectio/ios/BetterLectio/ReviewPromptCoordinator.swift
import Combine
import Foundation

/// Orchestrates soft rating pre-filter + App Store review after happy moments.
/// Mirrors Android `ReviewPromptCoordinator`.
@MainActor
final class ReviewPromptCoordinator: ObservableObject {
    static let shared = ReviewPromptCoordinator()

    @Published private(set) var softPromptVisible = false

    private let store: ReviewPromptStore
    private var launchRecordedThisProcess = false
    private var sessionStartedAt: Date?
    private var lastErrorAt: Date?
    private var pendingTask: Task<Void, Never>?
    private var shownTrigger: ReviewTrigger?
    private var responseHandled = false
    private var scheduleLoadedAttempted = false
    private var currentStudent: Student?
    private var blockers = Set<String>()

    private static let promptDelayNanoseconds: UInt64 = 1_800_000_000

    init(store: ReviewPromptStore = .shared) {
        self.store = store
    }

    /// Call once per authenticated process from the authenticated shell.
    /// Records a launch for the 8-opens / 14-days gate.
    func onAuthenticatedLaunch(student: Student) {
        currentStudent = student
        guard !student.isDemo else { return }
        guard !launchRecordedThisProcess else { return }
        launchRecordedThisProcess = true
        sessionStartedAt = Date()
        scheduleLoadedAttempted = false
        store.recordLaunch()
    }

    func reportRecentError() {
        lastErrorAt = Date()
    }

    func setExternalPromptBlocking(_ source: String, blocking: Bool) {
        if blocking {
            blockers.insert(source)
        } else {
            blockers.remove(source)
        }
    }

    /// Schedule-loaded is only attempted once per process (same as Android ViewModel gate).
    func maybePromptScheduleLoaded() {
        guard !scheduleLoadedAttempted else { return }
        scheduleLoadedAttempted = true
        maybePrompt(.scheduleLoaded)
    }

    /// After a happy moment — delays briefly, re-checks gates, then may show the soft sheet.
    func maybePrompt(_ trigger: ReviewTrigger) {
        guard let student = currentStudent, !student.isDemo else { return }
        guard !softPromptVisible else { return }

        pendingTask?.cancel()
        pendingTask = Task { [weak self] in
            try? await Task.sleep(nanoseconds: Self.promptDelayNanoseconds)
            guard let self, !Task.isCancelled else { return }
            guard !self.softPromptVisible else { return }
            if self.hasConflictingPrompt() {
                FeedbackLogBuffer.shared.record(
                    "Review prompt skipped: conflicting UI (\(trigger.analyticsName))"
                )
                return
            }
            guard ReviewEligibility.isEligible(
                store: self.store,
                sessionStartedAt: self.sessionStartedAt,
                lastErrorAt: self.lastErrorAt
            ) else {
                FeedbackLogBuffer.shared.record(
                    "Review prompt ineligible (\(trigger.analyticsName))"
                )
                return
            }
            self.shownTrigger = trigger
            self.responseHandled = false
            self.store.markSoftPromptShown()
            self.softPromptVisible = true
            Analytics.capture(
                "review_prompt_shown",
                properties: ["trigger": trigger.analyticsName]
            )
        }
    }

    func onPositive() {
        guard !responseHandled else {
            dismissSoftPromptUi()
            return
        }
        responseHandled = true
        let trigger = shownTrigger?.analyticsName ?? ""
        Analytics.capture("review_prompt_positive", properties: ["trigger": trigger])
        dismissSoftPromptUi()
        store.markStoreFlowRequested()
        Task {
            // Let the soft sheet finish dismissing before asking the system for a review.
            try? await Task.sleep(nanoseconds: 400_000_000)
            let launched = AppStoreReviewLauncher.request()
            Analytics.capture(
                "review_play_flow_requested",
                properties: [
                    "trigger": trigger,
                    "play_accepted": launched,
                ]
            )
        }
    }

    func onNegative() {
        guard !responseHandled else {
            dismissSoftPromptUi()
            return
        }
        responseHandled = true
        let trigger = shownTrigger?.analyticsName ?? ""
        Analytics.capture("review_prompt_negative", properties: ["trigger": trigger])
        dismissSoftPromptUi()
        if let student = currentStudent {
            FeedbackCoordinator.shared.present(for: student)
        }
    }

    func onDismissed() {
        guard !responseHandled else {
            dismissSoftPromptUi()
            return
        }
        responseHandled = true
        let trigger = shownTrigger?.analyticsName ?? ""
        Analytics.capture("review_prompt_dismissed", properties: ["trigger": trigger])
        dismissSoftPromptUi()
    }

    private func dismissSoftPromptUi() {
        softPromptVisible = false
        shownTrigger = nil
    }

    private func hasConflictingPrompt() -> Bool {
        if !blockers.isEmpty { return true }
        if ReferralCoordinator.shared.nudgeVisible { return true }
        if FeedbackCoordinator.shared.presentation != nil { return true }
        return false
    }
}

File: references/betterlectio/android/app/src/main/java/dk/betterlectio/android/feature/review/ReviewPromptCoordinator.kt
package dk.betterlectio.android.feature.review

import com.posthog.PostHog
import dk.betterlectio.android.core.lectio.session.SessionController
import dk.betterlectio.android.feature.feedback.FeedbackOpenRequests
import dk.betterlectio.android.feature.referral.ReferralCoordinator
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import timber.log.Timber
import java.time.Instant
import java.util.concurrent.atomic.AtomicBoolean
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Orchestrates soft rating pre-filter + Play review after happy moments.
 */
@Singleton
class ReviewPromptCoordinator @Inject constructor(
    private val store: ReviewPromptStore,
    private val eligibility: ReviewEligibility,
    private val playReviewLauncher: PlayReviewLauncher,
    private val session: SessionController,
    private val referralCoordinator: ReferralCoordinator,
    private val feedbackOpenRequests: FeedbackOpenRequests,
) {
    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main.immediate)
    private val launchRecordedThisProcess = AtomicBoolean(false)
    private var sessionStartedAt: Instant? = null
    private var lastErrorAt: Instant? = null
    private var pendingJob: Job? = null
    private var shownTrigger: ReviewTrigger? = null
    private val responseHandled = AtomicBoolean(false)

    /** Active soft-prompt blockers keyed by source (extension, feedback, …). */
    private val blockers = java.util.concurrent.ConcurrentHashMap.newKeySet<String>()

    private val _softPromptVisible = MutableStateFlow(false)
    val softPromptVisible: StateFlow<Boolean> = _softPromptVisible.asStateFlow()

    /**
     * Call once per authenticated process from the authenticated shell.
     * Records a launch for the 8-opens / 14-days gate.
     */
    fun onAuthenticatedLaunch() {
        if (!launchRecordedThisProcess.compareAndSet(false, true)) return
        sessionStartedAt = Instant.now()
        store.recordLaunch()
    }

    fun reportRecentError() {
        lastErrorAt = Instant.now()
    }

    fun setExternalPromptBlocking(source: String, blocking: Boolean) {
        if (blocking) blockers.add(source) else blockers.remove(source)
    }

    /**
     * After a happy moment — delays briefly, re-checks gates, then may show the soft sheet.
     */
    fun maybePrompt(trigger: ReviewTrigger) {
        val student = session.currentStudent ?: return
        if (student.isDemo) return
        if (_softPromptVisible.value) return

        pendingJob?.cancel()
        pendingJob = scope.launch {
            delay(PROMPT_DELAY_MS)
            if (_softPromptVisible.value) return@launch
            if (hasConflictingPrompt()) {
                Timber.d("Review prompt skipped: conflicting UI (%s)", trigger.analyticsName)
                return@launch
            }
            if (!eligibility.isEligible(
                    sessionStartedAt = sessionStartedAt,
                    lastErrorAt = lastErrorAt,
                )
            ) {
                Timber.d("Review prompt ineligible (%s)", trigger.analyticsName)
                return@launch
            }
            shownTrigger = trigger
            responseHandled.set(false)
            store.markSoftPromptShown()
            _softPromptVisible.value = true
            PostHog.capture(
                event = "review_prompt_shown",
                properties = mapOf("trigger" to trigger.analyticsName),
            )
        }
    }

    fun onPositive(activity: android.app.Activity) {
        if (!responseHandled.compareAndSet(false, true)) {
            dismissSoftPromptUi()
            return
        }
        val trigger = shownTrigger?.analyticsName.orEmpty()
        PostHog.capture(
            event = "review_prompt_positive",
            properties = mapOf("trigger" to trigger),
        )
        dismissSoftPromptUi()
        store.markPlayFlowRequested()
        playReviewLauncher.launch(activity) { launched ->
            PostHog.capture(
                event = "review_play_flow_requested",
                properties = mapOf(
                    "trigger" to trigger,
                    "play_accepted" to launched,
                ),
            )
        }
    }

    fun onNegative() {
        if (!responseHandled.compareAndSet(false, true)) {
            dismissSoftPromptUi()
            return
        }
        val trigger = shownTrigger?.analyticsName.orEmpty()
        PostHog.capture(
            event = "review_prompt_negative",
            properties = mapOf("trigger" to trigger),
        )
        dismissSoftPromptUi()
        feedbackOpenRequests.requestOpen()
    }

    fun onDismissed() {
        if (!responseHandled.compareAndSet(false, true)) {
            dismissSoftPromptUi()
            return
        }
        val trigger = shownTrigger?.analyticsName.orEmpty()
        PostHog.capture(
            event = "review_prompt_dismissed",
            properties = mapOf("trigger" to trigger),
        )
        dismissSoftPromptUi()
    }

    private fun dismissSoftPromptUi() {
        _softPromptVisible.value = false
        shownTrigger = null
    }

    private fun hasConflictingPrompt(): Boolean {
        if (blockers.isNotEmpty()) return true
        if (referralCoordinator.nudgeVisible.value) return true
        return false
    }

    companion object {
        private const val PROMPT_DELAY_MS = 1_800L
    }
}
